File: 84-largest-rectangle-in-histogram/84-largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        heights = [0] + heights + [0]
        stack = []
        res = 0
        for i in range(len(heights)):
            while stack and heights[stack[-1]] > heights[i]:
                j = stack.pop()
                res = max(res, heights[j] * (i-stack[-1]-1)) 
            stack.append(i)
        return res

File: 84-largest-rectangle-in-histogram/84-largest-rectangle-in-histogram.js
/**
 * @param {number[]} heights
 * @return {number}
 */
var largestRectangleArea = function(heights) {
  heights = [0, ...heights, 0];
  const stack = [];
  let res = 0;
  
  for (let i = 0; i < heights.length; i++) {
    while (stack.length && heights[stack.at(-1)] > heights[i]) {
      const j = stack.pop();
      res = Math.max(res, heights[j] * (i - stack.at(-1) - 1));
    }
    
    stack.push(i);
  }
  
  return res;
};